Adding a note to `gh search` docs to explain the usage of `--` to exclude certain results

Sukhpreet-s opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

Fixes #10480

This PR improves the `gh search` docs by explaining the usage of (`--`) when using **GitHub search syntax** as query to **exclude** results based on qualifiers. 

Docs for the following `gh` commands are updated:
- `gh search code`
- `gh search commit`
- `gh search issues`
- `gh search prs`
- `gh search repos`

### Looking for feedback / Things to still figure out:

1.  Any improvements in the language used for the note.

2. While testing the `--` behaviour, I found that the `--` syntax does not work as expected on Powershell. This might require an addition to the docs notifying this unexpected behaviour to the users.
**Command I used to test on Powershell:** `gh search issues distribution -- repo:rubyforgood/human-essentials -label:stale`
**Issue:** results included issues with the `stale` label

<hr/>

#### References used:
- https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests
    > Use a minus (hyphen) symbol to exclude results that match a qualifier. For example, to ignore issues created by the "octocat" user, you'd use -author:octocat in your search. Note that this does not work for [missing metadata qualifiers](https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests#search-by-missing-metadata).

- https://docs.github.com/en/search-github/getting-started-with-searching-on-github/understanding-the-search-syntax#exclude-results-that-match-a-qualifier
    > You can narrow down search results by excluding one or more subsets. To exclude all results that are matched by a qualifier, prefix the search qualifier with a hyphen (-).

- https://github.com/cli/cli/blob/2b89358543d1fe9c0be902992a05d6f6a2fe4b15/pkg/cmd/search/prs/prs.go#L34-L35

Sukhpreet-s committed:
docs(search): add note for exclusion search syntax

Sukhpreet-s commented:
> While testing the -- behaviour, I found that the -- syntax does not work as expected on Powershell. This might require an addition to the docs notifying this unexpected behaviour to the users.
> Command I used to test on Powershell: gh search issues distribution -- repo:rubyforgood/human-essentials -label:stale
> Issue: results included issues with the stale label

To investigate, I spoke with @BagToad, and we found that the request URL is not encoded correctly for certain query patterns.

> [!NOTE]
>  We used `$env:GH_DEBUG = "api"` to see detailed output.

## ❌ Command 1
```powershell
gh search issues -- distribution repo:rubyforgood/human-essentials -label:stale
```

**URL path:** 
```powershell
https://api.github.com/search/issues?page=1&per_page=30&q=distribution+repo%3Arubyforgood%2Fhuman-essentials+-label%3A+stale+type%3Aissue
```
**Problem:**
A stray `+` appears before `stale` in `-label%3A+stale`, causing incorrect results (includes issues with the stale label).

---

## ❌ Command 2: Full query in quotes

```powershell
gh search issues -- 'distribution repo:rubyforgood/human-essentials -label:stale'
```

**Request URL:**

```powershell
https://api.github.com/search/issues?page=1&per_page=30&q=distribution+repo%3A%22rubyforgood%2Fhuman-essentials+-label%3Astale%22+type%3Aissue
```

**Problem:**  
The entire query gets wrapped in quotes, leading to:

```powershell
Invalid search query "distribution repo:\"rubyforgood/human-essentials -label:stale\" type:issue".
```

---

## ✅ Command 3: Only exclusion part quoted

```powershell
gh search issues -- distribution repo:rubyforgood/human-essentials '-label:stale'
```

**Request URL:**

```powershell
https://api.github.com/search/issues?page=1&per_page=30&q=distribution+repo%3Arubyforgood%2Fhuman-essentials+-label%3Astale+type%3Aissue
```

**Result:**  
Works as expected — issues with the `stale` label are excluded.

Sukhpreet-s commented:
I noticed that the `--` separator works as expected in **Windows Command Prompt**, but the issue occurs specifically in **Windows PowerShell**.

For example, the following command:

`gh search issues -- distribution repo:rubyforgood/human-essentials -label:stale`

runs without problems in `cmd.exe`, but fails in PowerShell.

BagToad commented:
The unix stop parse character `--` is [implemented directly by cobra](https://github.com/spf13/cobra/blob/6dec1ae26659a130bdb4c985768d1853b0e1bc06/command.go#L688). So on powershell, you can leverage these with a combination of the powershell stop parse character `--%` + the `--` implemented by cobra:

```gh --% search issues --repo rubyforgood/human-essentials -- distribution -label:stale```

:point_up: the `--%` allows powershell to ignore everything after it, essentially handing the flag parsing to cobra. The `--` then functions the same as we’re used to in a unix shell.

BagToad committed:
Add help topic for search syntax in gh commands

Introduces a new help topic explaining how to use exclusion qualifiers in GitHub search syntax with gh commands, including platform-specific instructions for Unix-like systems and PowerShell. Provides links to relevant documentation for further reference.

Co-Authored-By: Sid <41968447+sukhpreet-s@users.noreply.github.com>

Sukhpreet-s committed:
docs(search): add reference to `gh help search`

Sukhpreet-s committed:
docs(help): rename `search-syntax` help topic

BagToad commented:
@andyfeller I've been working with @Sukhpreet-s on this one. Most of the conversation up till this point has been trying to understand how powershell works in this case. @Sukhpreet-s wanted to make sure we're doing right by powershell users by ensuring our documentation is complete for all platforms. 

We decided that the explanation was too meaty to be in each `search` subcommand and putting this information in the top-level `gh search --help` only might have discoverability issues. We decided to propose a help topic for this and reference _that_ from each search subcommand ensuring the linkage to the information is discoverable. 

Since I've been pretty in the weeds with this and I'm interested in your insight as the original issue author as to where you envisioned this might be documented, I requested your review. 

Another thing that came up was potentially moving this documentation to docs.github.com, but we decided to see what a help topic feels like first and get another opinion 🍻

andyfeller commented:
Thanks for the ping, @BagToad! 🙇 

1. I would suggest having `gh help search-syntax` content live in `gh search` long documentation
2. I wonder if we need something else for those new to working with terminals and shells to understand the various features that `gh` is designed to support
3. I would love to discuss ways to expand our help usage logic to allow content to show up in children without duplication

### Why I think `gh search` is the place to put this.

1. I don't think `gh help` should be commandset-specific as these help topics are cross cutting concerns

   Everything else in here is focused on broader topics like accessibility, environment variables, exit codes, JSON exporting, Go template formatting, etc.  Things that apply across all `gh` commands.

   Additionally, I support removing `gh help actions` topic as I'm unsure the value it provides over the Actions command group in `gh help`:

   ```shell
   $ gh help
   Work seamlessly with GitHub from the command line.
  
   ...
  
   GITHUB ACTIONS COMMANDS
     cache:         Manage GitHub Actions caches
     run:           View details about workflow runs
     workflow:      View details about GitHub Actions workflows
   ```

2. `gh help search-syntax` might confuse users looking for documentation on [search qualifier documentation](https://docs.github.com/en/search-github/searching-on-github)

   We have links in the respective `gh search` subcommands for documentation on search syntax.

3. I don't think help topics are more discoverable than `man ...` and `... --help` going up the commandset

   Though I say this because I've ascribed to the Unix philosophy for a while, but I'd say thats the expectation someone who is more comfortable with a terminal or shell would expect.

   This PR is already pointing users to run a different command for more information, which could be `gh search --help` and would be located more closely to the command set than buried under root help topics.

### How do we help users understand the terminal and shell behaviors `gh` is designed for?

I struggle with where the GitHub CLI provides broader resources for those new and unfamiliar with terminals and shells because the stop-parse delimiter is a broader topic than just `gh search`.

From [Bash documentation on built-in commands](https://www.gnu.org/software/bash/manual/bash.html#Shell-Builtin-Commands-1):

> Unless otherwise noted, each builtin command documented as accepting options preceded by `-` accepts `--` to signify the end of the options. The `:`, `true`, `false`, and `test`/`[` builtins do not accept options and do not treat `--` specially. The `exit`, `logout`, `return`, `break`, `continue`, `let`, and `shift` builtins accept and process arguments beginning with `-` without requiring `--`. Other builtins that accept arguments but are not specified as accepting options interpret arguments beginning with `-` as invalid options and require `--` to prevent this interpretation.

This is only one of several [shell expansion](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Shell-Expansions) features  `gh` is built around that users might be unfamiliar with:

- [Brace Expansion](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Brace-Expansion)
- [Tilde Expansion](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Tilde-Expansion)
- [Shell Parameter Expansion](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Shell-Parameter-Expansion)
- [Command Substitution](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Command-Substitution)
- [Arithmetic Expansion](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Arithmetic-Expansion)
- [Process Substitution](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Process-Substitution)
- [Word Splitting](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Word-Splitting)
- [Filename Expansion](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Filename-Expansion)
- [Quote Removal](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Quote-Removal)

### Can we make `--help` generation more manageable?

If there's a concern about managing documentation in multiple places, I'd love to soundboard ideas that could improve the management while allowing that content to appear where people look.

For example, could we enhance the help usage logic below to allow parent commands to define annotations that show up in their children's `--help` usage?

https://github.com/cli/cli/blob/7d70980d8329fe30385752515bf8131314f00931/pkg/cmd/root/help.go#L176-L194

BagToad commented:
👋 @Sukhpreet-s

@andyfeller and I discussed his thoughts a bit more and I wanted to summarize the next steps for this PR:

- Move the help topic content to `gh search --help`.
    - Andy and I both agree that this should go in `gh search --help`'s output instead. The content can stay the same. 

Do you feel comfortable making those changes? 🙂

Sukhpreet-s committed:
docs(search): move search syntax note to `gh search --help`

Sukhpreet-s commented:
@andyfeller  Thank you for the explanation on the decision to move the note to `search --help` command 💯 

Also, thank you @BagToad for simplifying the next steps ❤️✨ 

I have pushed the changes. 🚀

BagToad commented on pkg/cmd/search/search.go:
```diff
@@ -15,7 +16,33 @@ func NewCmdSearch(f *cmdutil.Factory) *cobra.Command {
+		Long:  heredoc.Docf(`
+			Search across all of GitHub.
+
+			Excluding search results that match a qualifier
```
I wish there was a good way to make this appear more like a heading, but that's not something that should be solved in this PR. Just making a comment about it.

BagToad commented on pkg/cmd/search/search.go:
```diff
@@ -15,7 +16,33 @@ func NewCmdSearch(f *cmdutil.Factory) *cobra.Command {
+			do not have the label "bug", you would use %[1]s-label:bug%[1]s as a search qualifier.
+
+			%[1]sgh%[1]s supports this syntax in %[1]sgh search%[1]s as well, but it requires extra
+			syntax to avoid the hyphen being interpreted as a command line flag because it begins with a hyphen.
```
```suggestion
			%[1]sgh%[1]s supports this syntax in %[1]sgh search%[1]s as well, but it requires extra
			command line arguments to avoid the hyphen being interpreted as a command line flag because it begins with a hyphen.
```

BagToad committed:
Update pkg/cmd/search/search.go

BagToad reviewed (approved):
LGTM - thanks for taking the time with me to work through this one! It definitely turned out to be more complicated. The fact that you **both** found and helped address the PowerShell difference was great, and no doubt will save someone a bunch of time in the future.

🍻

BagToad committed:
Fix linter

BagToad committed:
Merge branch 'trunk' into 10480-enhance-gh-search-docs
